Source/AGXUnreal/Private/Terrain/AGX_MovableTerrainComponent.cpp

```diff
@@ -0,0 +1,923 @@
+	: UProceduralMeshComponent(ObjectInitializer)
+{
+	PrimaryComponentTick.bCanEverTick = true;
+	SetCanEverAffectNavigation(false);
+}
+
+void UAGX_MovableTerrainComponent::BeginPlay()
+{
+	Super::BeginPlay();
+
+	// Create Heights
+	InitializeHeights();
+
+	// Create Native
+	CreateNative();
+	
+	// Create Mesh(s) and Tile(s)
+	InitializeMesh();
+
+	// Update Mesh each tick
+	if (UAGX_Simulation* Simulation = UAGX_Simulation::GetFrom(this))
+	{
+		PostStepForwardHandle =
+			FAGX_InternalDelegateAccessor::GetOnPostStepForwardInternal(*Simulation)
+				.AddLambda(
+					[this](double)
+					{
+						//Update CurrentHeights
+						NativeBarrier.GetHeights(CurrentHeights, true);
+
+						//Rebuild mesh if needed
+						auto DirtyHeights = NativeBarrier.GetModifiedVertices();
+						if (DirtyHeights.Num() > 0)
+							UpdateMesh(DirtyHeights);
+					});
+	}
+
+	// Init particles
+	if (bEnableParticleRendering)
+	{
+		if (!InitializeParticles())
+		{
+			UE_LOG(
+				LogAGX, Warning,
+				TEXT("InitializeParticles returned false in AGX_MovableTerrainComponent '%s'. "
+					 "Ensure the selected Niagara System is valid."),
+				*GetName());
+		}
+	}
+
+	// Unreal Collision
+	this->SetCollisionEnabled(AdditionalUnrealCollision);
+}
+
+void UAGX_MovableTerrainComponent::CreateNative()
+{
+	// Make sure OwningRigidBody is created, if there is one
+	UAGX_RigidBodyComponent* OwningRigidBody =
+		FAGX_ObjectUtilities::FindFirstAncestorOfType<UAGX_RigidBodyComponent>(*this);
+	if (OwningRigidBody)
+		OwningRigidBody->GetOrCreateNative();
+
+	// Resolution
+	FIntVector2 TerrainResolution = GetTerrainResolution();
+
+	// Allocate Native
+	NativeBarrier.AllocateNative(
+		TerrainResolution.X, TerrainResolution.Y, ElementSize, CurrentHeights, BedHeights);
+
+	// Attach to RigidBody
+	if (OwningRigidBody)
+		OwningRigidBody->GetNative()->AddTerrain(&NativeBarrier);
+
+	// Set transform
+	NativeBarrier.SetRotation(this->GetComponentQuat());
+	NativeBarrier.SetPosition(this->GetComponentLocation());
+
+	// Set Native Properties
+	NativeBarrier.SetCanCollide(bCanCollide);
+	NativeBarrier.AddCollisionGroups(CollisionGroups);
+	NativeBarrier.SetCreateParticles(bCreateParticles);
+	NativeBarrier.SetDeleteParticlesOutsideBounds(bDeleteParticlesOutsideBounds);
+	NativeBarrier.SetPenetrationForceVelocityScaling(PenetrationForceVelocityScaling);
+	NativeBarrier.SetMaximumParticleActivationVolume(MaximumParticleActivationVolume);
+
+	// Create/Add Shovels
+	CreateNativeShovels();
+
+	if (!UpdateNativeTerrainMaterial())
+	{
+		UE_LOG(
+			LogAGX, Warning,
+			TEXT("UpdateNativeTerrainMaterial returned false in AGX_MovableTerrainComponent '%s'. "
+				 "Ensure the selected Terrain Material is valid."),
+			*GetName());
+	}
+
+	if (!UpdateNativeShapeMaterial())
+	{
+		UE_LOG(
+			LogAGX, Warning,
+			TEXT("UpdateNativeShapeMaterial returned false in AGX_MovableTerrainComponent '%s'. "
+				 "Ensure the selected Shape Material is valid."),
+			*GetName());
```

Comment:
For all of the warnings/errors, it is good to write out the owner name as well. See for example RigidBodyComponent, and follow that pattern: 

```
UE_LOG(
		LogAGX, Warning,
		TEXT("Write Transform To Native called on Rigid Body Component '%s' in '%s' that does "
			 "not have a native. Doing nothing."),
		*GetName(), *GetLabelSafe(GetOwner()));
``` 

It is helpful because if a user have multiple MovableTerrains, they often end up with the same name "AGX_MovableTerrain", so it's difficult to know what Component gave the error/warning of the owner is not printed.